flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java

```diff
@@ -292,27 +313,34 @@ private MemorySegment requestMemorySegmentBlocking(int targetChannel) throws Int
 
 	@Nullable
 	private MemorySegment requestMemorySegment(int targetChannel) {
-		MemorySegment segment = null;
+		MemorySegment segment;
 		synchronized (availableMemorySegments) {
-			returnExcessMemorySegments();
-
-			if (availableMemorySegments.isEmpty()) {
-				segment = requestMemorySegmentFromGlobal();
+			if (isDestroyed) {
+				throw new IllegalStateException("Buffer pool is destroyed.");
 			}
-			// segment may have been released by buffer pool owner
-			if (segment == null) {
-				segment = availableMemorySegments.poll();
+
+			// target channel over quota; do not return a segment
+			if (targetChannel != UNKNOWN_CHANNEL && subpartitionBuffersCount[targetChannel] >= maxBuffersPerChannel) {
+				return null;
 			}
+
+			segment = availableMemorySegments.poll();
+
 			if (segment == null) {
-				availabilityHelper.resetUnavailable();
+				return null;
 			}
 
-			if (segment != null && targetChannel != UNKNOWN_CHANNEL) {
-				if (subpartitionBuffersCount[targetChannel]++ == maxBuffersPerChannel) {
+			if (targetChannel != UNKNOWN_CHANNEL) {
+				if (++subpartitionBuffersCount[targetChannel] == maxBuffersPerChannel) {
```

Comment:
`subpartitionBuffersCount[targetChannel]++` vs `++subpartitionBuffersCount[targetChannel]`, isn't it changing the semantic a bit?

Reply:
Yes, I should probably pull it out. Imho it was wrong before. For me valid `subpartitionBuffersCount`s are `[0, maxBuffersPerChannel]`, but it used to be `[0, maxBuffersPerChannel + 1]`. In the end, `maxBuffersPerChannel` is rather arbitrary, so it shouldn't have a any impact on production applications, but for me, it's also important to have a clear semantics.

Reply:
Yes, it would be better to move it to a separate commit. I agree that `[0, maxBuffersPerChannel]` makes more sense.